jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/SuspendResumeTest.java

```diff
@@ -0,0 +1,202 @@
+        clientSocket.session.getRemote().sendString("hello world");
+
+        assertThat(serverSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("suspend"));
+        assertNull(serverSocket.receivedMessages.poll(1, TimeUnit.SECONDS));
+
+        serverSocket.suspendToken.resume();
+        assertThat(serverSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("suspend"));
+        assertNull(serverSocket.receivedMessages.poll(1, TimeUnit.SECONDS));
+
+        serverSocket.suspendToken.resume();
+        assertThat(serverSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("hello world"));
+        assertNull(serverSocket.receivedMessages.poll(1, TimeUnit.SECONDS));
+
+        // make sure both sides are closed
+        clientSocket.session.close();
+        assertTrue(clientSocket.closed.await(5, TimeUnit.SECONDS));
+        assertTrue(serverSocket.closed.await(5, TimeUnit.SECONDS));
+
+        // check we closed normally
+        assertThat(clientSocket.closeCode, is(StatusCode.NORMAL));
+        assertThat(serverSocket.closeCode, is(StatusCode.NORMAL));
+    }
+
+    @Test
+    public void testExternalSuspend() throws Exception
+    {
+        URI uri = new URI("ws://localhost:"+connector.getLocalPort()+"/suspend");
+        EventSocket clientSocket = new EventSocket();
+        Future<Session> connect = client.connect(clientSocket, uri);
+        connect.get(5, TimeUnit.SECONDS);
+
+        // verify connection by sending a message from server to client
+        assertTrue(serverSocket.open.await(5, TimeUnit.SECONDS));
+        serverSocket.session.getRemote().sendString("verification");
+        assertThat(clientSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("verification"));
+
+        // suspend the client so that no read events occur
+        SuspendToken suspendToken = clientSocket.session.suspend();
+
+        // verify client can still send messages
+        clientSocket.session.getRemote().sendString("message-from-client");
+        assertThat(serverSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("message-from-client"));
+
+        // the message is not received as it is suspended
+        serverSocket.session.getRemote().sendString("message-from-server");
+        assertNull(clientSocket.receivedMessages.poll(2, TimeUnit.SECONDS));
+
+        // client should receive message after it resumes
+        suspendToken.resume();
+        assertThat(clientSocket.receivedMessages.poll(5, TimeUnit.SECONDS), is("message-from-server"));
+
+        // make sure both sides are closed
+        clientSocket.session.close();
+        assertTrue(clientSocket.closed.await(5, TimeUnit.SECONDS));
+        assertTrue(serverSocket.closed.await(5, TimeUnit.SECONDS));
+
+        // check we closed normally
+        assertThat(clientSocket.closeCode, is(StatusCode.NORMAL));
+        assertThat(serverSocket.closeCode, is(StatusCode.NORMAL));
+    }
+
+    @Test
+    public void testSuspendAfterClose() throws Exception
```

Comment:
Oooh, good point.   Good test.

Should we make it so that once the local Endpoint starts the websocket close handshake, the call to `suspend()` no longer works?
Can you think of any reason that we should, or shouldn't do that?

Reply:
I don't think we should do this, you could initiate the close handshake but still receive multiple data frames before receiving the close response, so you should still be able to suspend the delivery of those frames.